### ISSUE TITLE
NULL + 1 in _PyFunction_FastCallDict(), PyEval_EvalCodeEx()

### ISSUE THREAD

**Issue by serhiy-storchaka** [python/cpython#74825]
BPO | [30640](https://bugs.python.org/issue30640)
--- | :---
Nosy | @vstinner, @serhiy-storchaka, @ZackerySpytz
PRs | <li>python/cpython#2919</li><li>python/cpython#2964</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-09-08.01:19:03.544>
created_at = <Date 2017-06-12.14:37:59.165>
labels = ['interpreter-core', '3.7']
title = 'NULL + 1 in _PyFunction_FastCallDict(), PyEval_EvalCodeEx()'
updated_at = <Date 2017-09-08.01:19:03.543>
user = 'https://github.com/serhiy-storchaka'
```

bugs.python.org fields:
```python
activity = <Date 2017-09-08.01:19:03.543>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-09-08.01:19:03.544>
closer = 'vstinner'
components = ['Interpreter Core']
creation = <Date 2017-06-12.14:37:59.165>
creator = 'serhiy.storchaka'
dependencies = []
files = []
hgrepos = []
issue_num = 30640
keywords = []
message_count = 7.0
messages = ['295786', '295788', '295794', '299538', '299556', '301663', '301664']
nosy_count = 3.0
nosy_names = ['vstinner', 'serhiy.storchaka', 'ZackerySpytz']
pr_nums = ['2919', '2964']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue30640'
versions = ['Python 3.7']
```

</p></details>

**Comment by serhiy-storchaka** [comment]
In \_PyFunction_FastCallDict() if nk == 0, k is set to NULL. After that k + 1 is passed to \_PyEval_EvalCodeWithName(). NULL + 1 is an undefined behavior.

**Comment by vstinner** [comment]
Can you please show me the line doing NULL+1? I don't see it. Or do you mean the "k + 1" with k=NULL?

\_PyEval_EvalCodeWithName() ignores kwnames and kwargs when kwcount is zero. So I don't think that the value of NULL+1 matters here :-)

**Comment by serhiy-storchaka** [comment]
Yes, I mean the "k + 1" with k=NULL.

Even if \_PyEval_EvalCodeWithName() ignores that value, this still is an undefined behavior. For example the compiler can decide that since NULL+1 is not valid, then k never is NULL, and ignore the branch that sets k to NULL. It can also not generate the code in \_PyEval_EvalCodeWithName() for handling the case kwcount == 0.

**Comment by 2776f601-9573-4690-ab86-59139fdf3c89** [comment]
As mentioned in [PR 2919](https://github.com/python/cpython/pull/2919), this is an issue in PyEval_EvalCodeEx() as well.

**Comment by serhiy-storchaka** [comment]
New changeset c6ea8974e2d939223bfd6d64ee13ec89c090d2e0 by Serhiy Storchaka (Zackery Spytz) in branch 'master':
bpo-30640: Fix undefined behavior in \_PyFunction_FastCallDict() and PyEval_EvalCodeEx() (bpo-2919)
https://github.com/python/cpython/commit/c6ea8974e2d939223bfd6d64ee13ec89c090d2e0

**Comment by vstinner** [comment]
New changeset f032e9237aa7d43d21e0b04d685c36bddf7078c1 by Victor Stinner (Zackery Spytz) in branch '3.6':
[3.6] bpo-30640: Fix undefined behavior in \_PyFunction_FastCallDict() and PyEval_EvalCodeEx() (GH-2919) (bpo-2964)
https://github.com/python/cpython/commit/f032e9237aa7d43d21e0b04d685c36bddf7078c1

**Comment by vstinner** [comment]
It's now fixed in Python 3.6 and master (3.7).

FYI this issue was also detected by Coverity as CID 1415964.
